### PR TITLE
merge overlapping segments

### PR DESCRIPTION
**Pull request by macearl**
Merge segments which start and end time are less than 500ms apart.

i think this more a work in progress rather than finished.

I added this to the player listener as i don't really think the API class should merge normally distinct segments into each other, but I am also not sure if this is the best place for it.

I added some comments to a couple of lines, which I am not sure on yet.

### CODE REVIEW

File: resources/lib/player_listener.py [hunk 1]
```diff
@@ -31,6 +31,26 @@ def _sanity_check_segments(segments):  # type: (Iterable[SponsorSegment]) -> boo
     return True
 
 
+def merge_overlapping_segments(segments): # type: (Iterable[SponsorSegment]) -> Iterable[SponsorSegment]
+    mergedsegments = []
+    itersegments = iter(segments)
+    curseg = next(itersegments)
+
+    for seg in itersegments: # type: SponsorSegment
```
Comment (macearl) [inline resources/lib/player_listener.py:39]: Segments should be sorted already, check again anyway?

File: resources/lib/player_listener.py [hunk 2]
```diff
@@ -31,6 +31,26 @@ def _sanity_check_segments(segments):  # type: (Iterable[SponsorSegment]) -> boo
...
+        if seg.start - curseg.end <= 0.5:
+            logger.debug("%s: merging into previous segment", seg)
+            curseg = curseg._replace(end = seg.end)
+            logger.debug("%s: merged segment", curseg)
+        else:
+            mergedsegments.append(curseg)
+            curseg = seg
+
+    if not curseg in mergedsegments:
```
Comment (macearl) [inline resources/lib/player_listener.py:48]: the last segment is never already in the merged list, right? If so this check probably is not necessary/

File: resources/lib/player_listener.py [hunk 3]
```diff
@@ -31,6 +31,26 @@ def _sanity_check_segments(segments):  # type: (Iterable[SponsorSegment]) -> boo
...
 
+def merge_overlapping_segments(segments): # type: (Iterable[SponsorSegment]) -> Iterable[SponsorSegment]
+    mergedsegments = []
+    itersegments = iter(segments)
+    curseg = next(itersegments)
+
+    for seg in itersegments: # type: SponsorSegment
+        if seg.start - curseg.end <= 0.5:
+            logger.debug("%s: merging into previous segment", seg)
+            curseg = curseg._replace(end = seg.end)
```
Comment (macearl) [inline resources/lib/player_listener.py:42]: this just overwrites the old end time, should we also modify the other values, for example the category?
As we merge different types into each other there is no single category which works here

**Comment by macearl** [comment]
whoops, completely missed your commit: https://github.com/siku2/script.service.sponsorblock/commit/e33cf042c3e64eded2315efa7d3dab355e2235bf#

I'll test that tomorrow and see if it works as well, if yes this PR can probably be closed without merging

File: resources/lib/player_listener.py (hunk 3 above)
Comment (siku2) [reply resources/lib/player_listener.py:42]: That's exactly the reason I didn't go for this approach. I want to preserve the metadata so we can report the segments (requires the uuid) and ideally show the user which categories were skipped (although I never got to this).
I considered adding a `MergedSponsorSegment` class that contains a list of segments and exposes the total start / end time, but the approach I ended up taking took less time to implement :P

**Comment by macearl** [comment]
closing this as another solution for this was implemented in https://github.com/siku2/script.service.sponsorblock/commit/e33cf042c3e64eded2315efa7d3dab355e2235bf

**Pull request closed without merging**